### ISSUE TITLE
Resend Verification Email

### ISSUE THREAD

**Issue by LukaJaj** [ory/kratos#2892]
### Preflight checklist

- [X] I could not find a solution in the existing issues, docs, nor discussions.
- [X] I agree to follow this project's [Code of Conduct](https://github.com/ory/kratos/blob/master/CODE_OF_CONDUCT.md).
- [X] I have read and am following this repository's [Contribution Guidelines](https://github.com/ory/kratos/blob/master/CONTRIBUTING.md).
- [ ] This issue affects my [Ory Network](https://www.ory.sh/) project.
- [ ] I have joined the [Ory Community Slack](https://slack.ory.sh).
- [ ] I am signed up to the [Ory Security Patch Newsletter](https://ory.us10.list-manage.com/subscribe?u=ffb1a878e4ec6c0ed312a3480&id=f605a41b53).

### Describe your problem

I'm trying to figure out how it would be possible to resend verification email to users who haven't verified their accounts but I was not able to find this functionality with current version of kratos

### Describe your ideal solution

it would be nice to include this functionality with verification flow to resend reminder email after some time.

### Workarounds or alternatives

build reminder link but in db tokens are hmac decrypted 

### Version

v0.10.1

**Comment by aeneasr** [comment]
That's a cool idea! It should also work for the new `code` flows.

**Comment by Sytten** [comment]
Agreed since there is not even a way to force kratos to mark the user as verified easily

**Comment by winston0410** [comment]
I think this is an issue that needs to be look into, as the current API does not provide any workaround for resending a email, for both recovery and verification flow. I tried to force a resend by removing the `flow` id, init a new one and do the first call programmatically(bring the state of the flow from `choose_method` to `sent_email`), but I will have to build extra features for handling the cooldown for that resend email button. It will be great if all these can be provided out of the box by Ory Kratos.

I imagine we can do this in the config:

```yaml
    verification:
        resend:
            cooldown: '5m'
```

and resend email with a `POST /self-service/verification/resend`

@aeneasr what do you think about this suggestion?

**Comment by Styren** [comment]
Any updates on this? If our users don't verify their email the first time they receive the mail then they are unable to signup at all. Are there any recommended workarounds?

**Comment by ChristianSch** [comment]
This is quite an oversight imho. I just launched a product using kratos today and  every single user had the problem that they couldn't click the link ... because the default template doesn't format it as such, even tho it says "click the following link". So what happened? They waited and the link was invalid. Now I have no idea how to resend the link because ory.sh has no way of doing that. Nor anything I've implemented.

**Comment by Sytten** [comment]
@winston0410 Did you ever get a response on that?

**Comment by Sytten** [comment]
@winston0410 Are you still interested in implementing it? I asked for some feedback on slack so we can make sure this design is validated.

**Comment by aeneasr** [comment]
I'd like to start with the end in mind. So what are the use cases where this can be used? I can come up with:

1. User triggers another verification email: For example when an app shows a banner "Please verify your email" with a link the user clicks.
2. Verification email is re-sent automatically after X days after registration, because the user is not yet verified

In my view, use case 1 is already possible using the verification flow. And use case 2 can be solved with the example in 1.

Furthermore, 2 is automted so not user-initiated. The chance that the user is in front of the email program at that exact time (before the code times out), is very unlikely.

What other use cases do we have for this that I missed?

**Comment by Sytten** [comment]
I think case 1 is what is important here?
The original use case of the issue as I understand it was:
1. User signs up with email/password
2. Kratos sends the verification email
3. A day passes
4. Users click on link and it's now expired

Similar issue probably occurs with a code, though I have not tried it myself. 
I see two possible problems:
1. verification link / code has expired (defined by `methods->link->config->lifespan`)
2. verification flow has expired (defined by `flows->verification->lifespan`)

Problem 1 can be solved by resolving problem 2 (aka you can't create a new link/code, but you can create a new flow). Though it is a weird behaviour and I remember starting a discussion about that a couple months ago.

Problem 2 is where I would focus the attention.
If I understand correctly, it is possible to start a new verification flow for that newly registered user?
If so what would be the process? I was imagining something like:
1. User logs in
2. Frontend shows: `enter the verification code`
3. User enters the expired code and receives an error
4. Frontend displays a button to get a new code
5. This triggers some backend call that creates a new verification flow and sends the email
6. Users enters the new code

Now we would probably want to have a de-bounce period between emails to avoid spam and we probably want to invalidate old codes/links when a new one is sent. I don't know about the internals of kratos enough to see if that is hard to do if we rely on creating a new verification flow each time.

**Comment by Benehiko** [comment]
Hi @Sytten 

It is possible to create a new verification flow, and can be created multiple times over in cases where the user navigates away or just exists the app. Kratos can also inform your UI that the user has not been verified after login.

In the verification flow the user would need to provide their email address. In a custom UI you can also just pre-fill the email. Once the user has created the verification flow, we provide a `resend` button which the user can submit to retrieve a new code. 
![image](https://github.com/ory/kratos/assets/18033717/d7198b9f-7659-4772-bb0b-e46a4852a52d)

The previous code would be expired by the resend request. All of these flows are client-side initiated and requires no admin interaction. It is also not needed to have an automated resend process here since the user could always be prompted by the UI to complete the verification process.

https://www.ory.sh/docs/kratos/self-service/flows/verify-email-account-activation

**Comment by Yiling-J** [comment]
I want to show a resend button on my app's signup page, and active it after 1 minutes. After take a look API, I think the name [Complete Verification Flow](https://www.ory.sh/docs/reference/api#tag/frontend/operation/updateVerificationFlow) is a bit misleading, you can use this API to resend the email, just exclude the code field. As the description of email field say:
```
The email address to verify

If the email belongs to a valid account, a verifiation email will be sent.

If you want to notify the email address if the account does not exist, see the [notify_unknown_recipients flag](https://www.ory.sh/docs/kratos/self-service/flows/verify-email-account-activation#attempted-verification-notifications)

If a code was already sent, including this field in the payload will invalidate the sent code and re-send a new code.
```

**Comment by Yiling-J** [comment]
I think the "Complete Verification Flow" thing actually not working though, this line https://github.com/ory/kratos/blob/4ce179a8c4cc431d3539ea1dc96c806b9f5d3beb/selfservice/strategy/code/strategy_verification.go#L190C4-L190C4 reject requests without email, but from the doc:
`If you want to submit a code, use this field, but make sure to not include the email field, as well.`

**Comment by Sytten** [comment]
No stale though I am not sure if it was fixed

**Comment by nis6** [comment]
When this user tries to register with the same email, I show id already exists using kratos/identities endpoint, and when they try to login, (based on the error from login flow) using the same email id I redirect them to the verification page along with the email id and auto-initiate the verification flow with this email id, so after email and password they just have to fill in the otp and verify themselves.

**Comment by Sytten** [comment]
Is this fixed? I dont remember honestly
